shimmy/meltingpot_compatibility.py

```diff
@@ -0,0 +1,193 @@
+from pettingzoo.utils.env import ObsDict, ActionDict
+
+import shimmy.utils.meltingpot as utils
+from meltingpot.python import substrate
+
+
+class MeltingPotCompatibilityV0(ParallelEnv, EzPickle):
+    """This compatibility wrapper converts a meltingpot substrate into a pettingzoo environment.
+
+    Melting Pot is a research tool developed to facilitate work on multi-agent artificial intelligence.
+    It assesses generalization to novel social situations involving both familiar and unfamiliar individuals,
+    and has been designed to test a broad range of social interactions such as: cooperation, competition,
+    deception, reciprocation, trust, stubbornness and so on.
+    Melting Pot offers researchers a set of over 50 multi-agent reinforcement learning substrates (multi-agent games)
+    on which to train agents, and over 256 unique test scenarios on which to evaluate these trained agents.
+    """
+
+    metadata = {'render_modes': ['human', 'rgb_array']}
+
+    PLAYER_STR_FORMAT = 'player_{index}'
+    MAX_CYCLES = 1000
+
+    def __init__(
+            self,
+            substrate_name: str,
+            render_mode: str | None,
+            max_cycles: int = MAX_CYCLES
+    ):
+        """Wrapper that converts a openspiel environment into a pettingzoo environment.
+
+        Args:
+            substrate_name (str): name of meltingpot substrate to load
+            render_mode (Optional[str]): render_mode
+            max_cycles (Optional[int]): maximum number of cycles (steps) before termination
+        """
+        # Load substrate from pickle
+        self.substrate_name = substrate_name
+        self.render_mode = render_mode
+        self.player_roles = substrate.get_config(self.substrate_name).default_player_roles
+        self.max_cycles = max_cycles
+        self.env_config = {"substrate": self.substrate_name, "roles": self.player_roles}
+        EzPickle.__init__(self, self.render_mode, self.env_config, self.max_cycles)
+
+        # Build substrate
+        self.env_config = config_dict.ConfigDict(self.env_config)
+        self._env = substrate.build(self.env_config['substrate'], roles=self.env_config['roles'])
+
+        self.state_space = utils.spec_to_space(
+            self._env.observation_spec()[0]['WORLD.RGB'])  # type: ignore
+
+        # Set agents
+        self._num_players = len(self._env.observation_spec())
+        self.possible_agents = [
+            self.PLAYER_STR_FORMAT.format(index=index)
+            for index in range(self._num_players)
+        ]
+        self.agents = [agent for agent in self.possible_agents]
+
+    def state_space(self) -> gymnasium.spaces.Space:
+        """observation_space.
+
+        Get the state space from the underlying meltingpot substrate.
+
+        Returns:
+            state_space: spaces.Space
+        """
+
+        state_space = utils.spec_to_space(
+            self._env.observation_spec()[0]['WORLD.RGB'])
+        return state_space
```

Comment:
Do you mean to make it in the init constructor rather than a separate method? I had it that way originally but saw the action space and observation space in most other examples were separate methods so thought it might be more readable that way. Although those are functions with agent id as input, whereas this is independent of agent. Wasn’t sure if it was necessary at all because a lot of examples don’t have state space included.

Reply:
It is just confusing as the `observation-spec` doesn't change and you already define `self.state_space` which is equivalent to this function. So just use the variable that already exists or remove the variable